codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java

```diff
@@ -1143,86 +825,44 @@ public long getTimeMillisAndRemove(T name, long defaultValue) {
-        }
-
-        return true;
+        DefaultHeaders<T> other = (DefaultHeaders<T>) o;
+        return size() == other.size() && map.equals(other.map);
     }
 
     /**
-     * Compare two lists using the {@code comparator} for all comparisons (not using the equals() operator)
-     * @param lhs Left hand side
-     * @param rhs Right hand side
-     * @param comparator Comparator which will be used for all comparisons (equals() on objects will not be used)
-     * @return True if {@code lhs} == {@code rhs} according to {@code comparator}. False otherwise.
+     * This method is purposfully kept simple and returns {@link #size()} as the hash code.
+     * There are two compelling reasons for keeping {@link #hashCode()} simple:
+     *   1) It's difficult to get it right as the hash code mostly depends on the {@link Map}
+     *      implementation. Simply using {@link Map#hashCode()} doesn't work as for
+     *      example {@link TreeMap#hashCode()} does not fullfill the contract between {@link Object#hashCode()} and
+     *      {@link Object#equals(Object)} when it's used with a {@link Comparator} that is not consistent with
+     *      {@link Object#equals(Object)}.
+     *   2) The {@link #hashCode()} function doesn't appear to be important for {@link Headers} implementations. It's
+     *      soley there because we overode {@link Object#equals(Object)}, which makes it easier to use {@link Headers}
+     *      in tests, but also has little practical use outside of tests.
      */
-    private static <T> boolean equals(List<T> lhs, List<T> rhs, Comparator<? super T> comparator) {
-        final int lhsSize = lhs.size();
-        if (lhsSize != rhs.size()) {
-            return false;
-        }
-
-        // Don't use a TreeSet to do the comparison.  We want to force the comparator
-        // to be used instead of the object's equals()
-        Collections.sort(lhs, comparator);
-        Collections.sort(rhs, comparator);
-        for (int i = 0; i < lhsSize; ++i) {
-            if (comparator.compare(lhs.get(i), rhs.get(i)) != 0) {
-                return false;
-            }
-        }
-        return true;
-    }
-
     @Override
     public int hashCode() {
-        int result = 1;
-        for (T name : names()) {
-            result = HASH_CODE_PRIME * result + name.hashCode();
-            List<T> values = getAll(name);
-            Collections.sort(values, valueComparator);
-            for (int i = 0; i < values.size(); ++i) {
-                result = HASH_CODE_PRIME * result + hashCodeGenerator.generateHashCode(values.get(i));
-            }
-        }
-        return result;
+        return size();
     }
 
     @Override
     public String toString() {
         StringBuilder builder = new StringBuilder(getClass().getSimpleName()).append('[');
+        String separator = "";
         for (T name : names()) {
             List<T> values = getAll(name);
-            Collections.sort(values, valueComparator);
```

Comment:
Understood.  I think this is a left over from development when the hierarchy was unified...made it much easier to look at headers and see the differences.  I wonder if it may still be useful to provide some sort of utility method which does the same...